### ISSUE TITLE
Feature suggestion: filter by module

### ISSUE THREAD

**Issue by paulochf** [coursera-dl/coursera-dl#569]
I used coursera-dl now and I found that a week from one course I want to download has the following structure

```
Processing module  simple-linear-regression
Processing section     regression-fundamentals
Processing lecture         slides-presented-in-this-module (supplement)
Processing lecture         a-case-study-in-predicting-house-prices (lecture)
Processing lecture         regression-fundamentals-data-model (lecture)
Processing lecture         regression-fundamentals-the-task (lecture)
Processing lecture         regression-ml-block-diagram (lecture)
Processing section     the-simple-linear-regression-model-its-use-and-interpretation
Processing lecture         the-simple-linear-regression-model (lecture)
Processing lecture         the-cost-of-using-a-given-line (lecture)
Processing lecture         using-the-fitted-line (lecture)
Processing lecture         interpreting-the-fitted-line (lecture)
Processing section     an-aside-on-optimization-one-dimensional-objectives
Processing lecture         defining-our-least-squares-optimization-objective (lecture)
Processing lecture         finding-maxima-or-minima-analytically (lecture)
Processing lecture         maximizing-a-1d-function-a-worked-example (lecture)
Processing lecture         finding-the-max-via-hill-climbing (lecture)
Processing lecture         finding-the-min-via-hill-descent (lecture)
Processing lecture         choosing-stepsize-and-convergence-criteria (lecture)
Processing section     an-aside-on-optimization-multidimensional-objectives
Processing lecture         gradients-derivatives-in-multiple-dimensions (lecture)
Processing lecture         gradient-descent-multidimensional-hill-descent (lecture)
Processing section     finding-the-least-squares-line
Processing lecture         computing-the-gradient-of-rss (lecture)
Processing lecture         approach-1-closed-form-solution (lecture)
Processing lecture         optional-reading-worked-out-example-for-closed-form-solution (supplement)
```

As I would like to download just the module above and another one, there could be a feature to pass which ones I would want to download, maybe giving the number indexes or the module title.

I just wanted to pass by and leave the suggestion. :smile:

Thanks!

**Comment by balta2ar** [comment]
There is no module filter option, but I think you can take advantage of section filter option instead if you really need to:
https://github.com/coursera-dl/coursera-dl/blob/master/coursera/commandline.py#L121

**Comment by paulochf** [comment]
Yeah, I did look in the section filter's code, then looked for any filter there, and found nothing. Also, section filter is not what I needed, so that's why I wrote the suggestion.

**Comment by balta2ar** [comment]
Could you please explain your motivation behind that? Why would you want to download one section and skip others?

**Comment by paulochf** [comment]
Did you mean "section"? I'm talking about "module" here. But I can answer for both cases.

For modules, the use case is that I have a friend that wanted to know about gradient descent. I recommended the above course, and there's two modules where this topic is explained. Given that, a module filter would be useful, so I could download only the modules instead of whole course.

For sections, I could skip the first ones, because in general they're introductory and would not be interesting for me.

For lectures, even finding that filtering here could kill the context of the topic, I understand that could be useful someway, as filters for the superior levels exist.


Personally, I find a module filter more useful than a lecture filter, and notice that I find the latter has it's use.

I could try to make a PR, but I preferred to open the suggestion here and give the opportunity to someone with more free time than me to do that, instead keeping the idea with me and this wonderful package doesn't get improved.

**Comment by rdrey** [comment]
Yeah, in the course I'm doing each week is a module, with two assignments due at the end of the module. I don't need to download far in advance, but would like to grab this week & the next week easily (by module filter).

**Comment by balta2ar** [comment]
> so I could download only the modules instead of whole course

Can you download the whole course and then remove the useless parts? If you're not short on bandwidth, that could be a workaround.

> I don't need to download far in advance

But you will need the whole course anyway, why not download it right now?

Generally, I am against filters as to me it as an unnecessary over-complication of the script. If your traffic costs a fortune, that's a more reasonable argument. In that case, you can run the script as is, it will generate syllabus in a JSON file. Then you can edit it in your editor and cut off the parts you don't need. Then you can use `--cache-syllabus` option to tell coursera-dl to reuse syllabus from the disk. It should download only the parts that were left in the file.

**Comment by rdrey** [comment]
Editing the syllabus sounds like a reasonable workaround, thanks.

**Comment by balta2ar** [comment]
> you can run the script as is, it will generate syllabus in a JSON file

Sorry, I missed the important part in that sentence. Just to be clear: run the script, **wait until it parses and saves syllabus, and terminate it manually** (Ctrl-C) when it starts downloading materials, before it downloads too much. Then edit the syllabus, add the flag and run again. There is no flag to exit after saving just syllabus so this approach is a bit hacky.
